dev/NumberBox/NumberBoxParser.cpp

```diff
@@ -0,0 +1,301 @@
+            {
+                while (!operatorStack.empty() && operatorStack.top().Char != L'(')
+                {
+                    // Pop operators onto output until we reach a left paren
+                    postfixTokens.push_back(operatorStack.top());
+                    operatorStack.pop();
+                }
+
+                if (operatorStack.empty())
+                {
+                    // Broken parenthesis
+                    error = true;
+                    break;
+                }
+
+                // Pop left paren and discard
+                operatorStack.pop();
+            }
+        }
+    }
+
+    // Pop all remaining operators.
+    while (!operatorStack.empty())
+    {
+        if (operatorStack.top().Type == MathTokenType::Parenthesis)
+        {
+            // Broken parenthesis
+            error = true;
+            break;
+        }
+
+        postfixTokens.push_back(operatorStack.top());
+        operatorStack.pop();
+    }
+
+    if (error)
+    {
+        return {};
+    }
+    return postfixTokens;
+}
+
+winrt::IReference<double> NumberBoxParser::ComputePostfixExpression(std::vector<MathToken> tokens)
+{
+    winrt::IReference<double> value = nullptr;
+    std::stack<double> stack;
+    bool error = false;
+
+    for (auto const token : tokens)
+    {
+        if (error)
+        {
+            break;
+        }
+
+        if (token.Type == MathTokenType::Operator)
+        {
+            // There has to be at least two values on the stack to apply
+            if (stack.size() < 2)
+            {
+                error = true;
+                break;
+            }
+
+            const auto op1 = stack.top();
+            stack.pop();
+
+            const auto op2 = stack.top();
+            stack.pop();
+
+            double result;
+
+            switch (token.Char)
+            {
+                case L'-':
+                    result = op2 - op1;
+                    break;
+
+                case L'+':
+                    result = op1 + op2;
+                    break;
+
+                case L'*':
+                    result = op1 * op2;
+                    break;
+
+                case L'/':
+                    if (op1 == 0)
+                    {
+                        error = true;
+                        value = NAN;
+                    }
+                    else
+                    {
+                        result = op2 / op1;
+                    }
+                    break;
+
+                case L'^':
+                    result = std::pow(op2, op1);
+                    break;
+
+                default:
+                    error = true;
+                    break;
+            }
+
+            stack.push(result);
+        }
+        else if (token.Type == MathTokenType::Numeric)
+        {
+            stack.push(token.Value);
+        }
+    }
+
+    // If there is more than one number on the stack, we didn't have enough operations, which is also an error.
```

Comment:
>which is also an error [](start = 88, length = 22)

If this is an error should we be returning the top of the stack or should we be signaling an error in some way?

Reply:
It is returning the error case, where value will be null.

Reply:
Oh i see now, here I think 
```
if(!error && stack.size() == 1)
{
    return stack.top()
}
else
{
    return nullptr;
}
```
would be better, you can delete the value local that way as well.

---
In reply to: [345450269](https://github.com/microsoft/microsoft-ui-xaml/pull/1602#discussion_r345450269) [](ancestors = 345450269)

Reply:
Fixed while removing the error variable.